[SINT-3775] Use CI Identities IAM Role in Windows Jobs

cedricvanrompay-datadog opened the pull request:
### What does this PR do?

Invoke `./ci-identities-gitlab-job-client.exe assume-role` at the beginning of each Windows CI job. This performs an `AssumeRole` targeting the CI Identity this CI job was given [(see the CI Identities user guide)](https://datadoghq.atlassian.net/wiki/spaces/SECENG/pages/5324145720/Using+CI+Identities) and stores the credentials in the "AWS shared credentials" file.

Note that:
* for now we have to download the client in the CI job, but later the client will be added to the image of the CI job to speed things up
* many Windows CI jobs run docker containers, we have to also invoke `./ci-identities-gitlab-job-client.exe assume-role` in them.

Make Omnibus use the CI Identity IAM role of the CI job to interact with AWS.

### Motivation

Beginning the onboarding to the "CI Identities" platform for more fine-grained permissions in CI jobs.

### Describe how you validated your changes

### Additional Notes

* [ ] backport this to the release branches (including 6.53.x)

cedricvanrompay-datadog committed:
test setting AWS env vars in omnibus invocation

chouetz commented:
Just for the record, I think we should not forget to backport this to the release branches (including `6.53.x`) once this is ready.

cedricvanrompay-datadog committed:
use ci-identities-gitlab-job-client

cedricvanrompay-datadog committed:
try fixing path to CI Identities client

cedricvanrompay-datadog committed:
fix usage of CI ID client

cedricvanrompay-datadog committed:
add GitLab ID token

cedricvanrompay-datadog committed:
add GitLab ID token directly to job

cedricvanrompay-datadog committed:
pass CI ID token through Docker

cedricvanrompay-datadog committed:
try again direct id token spec

cedricvanrompay-datadog committed:
debug prints

cedricvanrompay-datadog committed:
try fixing Build-AgentPackages.ps1

cedricvanrompay-datadog committed:
pass env through env variable

cedricvanrompay-datadog committed:
remove debug prints

cedricvanrompay-datadog committed:
fix ZIP jobs

cedricvanrompay-datadog committed:
injecting fake AWS creds for debugging

cedricvanrompay-datadog committed:
assume CI ID role in common.ps1

cedricvanrompay-datadog committed:
try fixing invocation of CI ID client

cedricvanrompay-datadog committed:
also fix invocation of client in before_script

cedricvanrompay-datadog committed:
stop wrapping omnibus execution with CI ID client

not needed as the CI ID client should have been called earlier
and wrote the creds to a file

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

cedricvanrompay-datadog committed:
use env var AWS_SHARED_CREDENTIALS_FILE

avoids side-effects
requires CI ID client version > 0.0.2

cedricvanrompay-datadog committed:
add comment about AWS_SHARED_CREDENTIALS_FILE

cedricvanrompay-datadog commented:
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable DEPLOY_AGENT=true --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

cedricvanrompay-datadog committed:
append job ID to STS session name

cedricvanrompay-datadog commented on .gitlab/trigger_distribution/conditions.yml:
```diff
@@ -171,6 +171,15 @@
     OVERRIDE_GIT_STRATEGY: "clone"
+    # Avoid using ~/.aws/credentials to avoid conflicts with other CI jobs
+    # this env var is understood by both ci-identities-gitlab-job-client.exe and aws.exe
+    AWS_SHARED_CREDENTIALS_FILE: ${CI_PROJECT_DIR}\.aws\credentials-by-job-id\${CI_JOB_ID}
```
No need, in the container the credentials file can located in the default location. Note that in the container we re-execute `./ci-identities-gitlab-job-client.exe assume-role`

cedricvanrompay-datadog commented on tasks/winbuildscripts/common.ps1:
```diff
@@ -342,3 +342,16 @@ function Invoke-BuildScript {
     }
 }
+
+Write-Host "Downloading CI identity client..."
```
Yes, I plan to move it to a function.

Good point, we should not run in on a laptop.

cedricvanrompay-datadog commented:
We do see CloudTrail logs matching `@requestParameters.roleSessionName:datadog-agent-builder-job-*` but the session name is exactly `datadog-agent-builder-job-`, there is no job ID. Looks like this code does not have `CI_JOB_ID` in its environment.

cedricvanrompay-datadog committed:
pass CI_JOB_ID env var to omnibus

cedricvanrompay-datadog commented:
OK, now we see the job IDs in the session name. This will be very helpful.

cedricvanrompay-datadog committed:
handle missing CI_JOB_ID in env

cedricvanrompay-datadog committed:
pass CI JOB ID to docker container

cedricvanrompay-datadog committed:
disable s3_sts_creds_instance_profile in omnibus

cedricvanrompay-datadog committed:
use credentials file in omnibus cache bucket

cedricvanrompay-datadog committed:
fix creds fiel path name

cedricvanrompay-datadog committed:
put an absolute path in s3_credentials_file_path

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

aiuto reviewed (commented):
Can you add `-e CI_JOB_ID=${CI_JOB_ID}` (and whatever else you think we need) here too:
https://github.com/DataDog/datadog-agent/blob/a39dfcb3ce9ecd3f57f9805dcf6a01d9ab25e725/.gitlab/bazel/defs.yaml#L63

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

cedricvanrompay-datadog committed:
pass GitLab ID token to container in .bazel:runner:windows-amd64

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

cedricvanrompay-datadog committed:
use newer version of the client

cedricvanrompay-datadog committed:
move CI ID init to a PowerShell fn in common.ps1

cedricvanrompay-datadog committed:
pass GitLab token to container in .integration_tests_windows_base

cedricvanrompay-datadog committed:
pass GitLab token to container in .lint_windows_base

cedricvanrompay-datadog committed:
pass GitLab token to container in .tests_windows_base

clarkb7 commented on tasks/omnibus.py:
```diff
@@ -50,6 +50,11 @@ def omnibus_run_task(ctx, task, target_project, base_dir, env, log_level="info",
+        if 'CI_JOB_ID' in os.environ:
+            env['CI_JOB_ID'] = os.environ.get('CI_JOB_ID')
+        else:
+            print("WARNING: CI_JOB_ID not present in environment")
```
should we only display this warning when running in CI?

cedricvanrompay-datadog committed:
do not warn about missing CI job ID if not in CI

cedricvanrompay-datadog committed:
pass CI job ID in bazel jobs as well

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

cedricvanrompay-datadog committed:
stop downloading client at runtime

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

cedricvanrompay-datadog committed:
use new Windows build image

from https://github.com/DataDog/datadog-agent-buildimages/pull/1065

cedricvanrompay-datadog committed:
update path to CI ID client

cedricvanrompay-datadog committed:
revert path change for invocation in Windows runner

In the Windows CI job runner, the client is at C:\ci-identities-gitlab-job-client.exe
while in the WINBUILDIMAGE it is at C:\devtools\ci-identities-gitlab-job-client.exe

cedricvanrompay-datadog committed:
use CI Identities in windows 2019 and 2025 jobs as well

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

cedricvanrompay-datadog committed:
update Windows CI image

KevinFairise2 reviewed (approved):
Thank you!

cedricvanrompay-datadog committed:
Merge branch 'main' into cedric.vanrompay/sint-3775-omnibus-use-ci-identity

rdesgroppes commented on tasks/omnibus.py:
```diff
@@ -50,6 +50,11 @@ def omnibus_run_task(ctx, task, target_project, base_dir, env, log_level="info",
+        if 'CI_JOB_ID' in os.environ:
+            env['CI_JOB_ID'] = os.environ.get('CI_JOB_ID')
+        elif os.environ.get("CI"):
+            print("WARNING: CI_JOB_ID not present in environment")
```
indeed, please consider adding `CI_JOB_ID` to `ENV_PASSHTROUGH` instead of introducing a special case.

cedricvanrompay-datadog committed:
stop forwarding CI_JOB_ID

since 83a2932 we don't need to forward it anymore

cedricvanrompay-datadog committed:
set AWS_SHARED_CREDENTIALS_FILE in .windows_docker_2025

all Windows CI jobs need it

rdesgroppes reviewed (approved):
LGTM, thanks for your patience!

cedricvanrompay-datadog commented:
/merge
